### PR TITLE
Extract: Ability to delete extracted events

### PR DESCRIPTION
**Pull request by PopDaph**
This PRs add a button "Delete data" on the list of extracted events. 

Capture d’écran 2023-07-25 à 09 33 15

### CODE REVIEW

File: front/lib/api/extract.ts [hunk 1]
```diff
@@ -154,3 +154,37 @@ export async function getExtractedEvents(
...
   });
 }
+
+export async function deleteExtractedEvent(
+  auth: Authenticator,
+  id: string
+): Promise<boolean> {
+  const owner = auth.workspace();
+  if (!owner) {
+    throw "extracted_event_auth_error";
```
Comment (spolu) [inline front/lib/api/extract.ts:164]: We don't throw error types, we'd rather have a sentence for now

Comment (spolu) [reply front/lib/api/extract.ts:164]: (to ensure consistency with the rest of the files in `lib/api`)

File: front/lib/api/extract.ts [hunk 2]
```diff
@@ -154,3 +154,37 @@ export async function getExtractedEvents(
...
+  }
+
+  const event = await ExtractedEvent.findOne({
+    where: {
+      id: id,
+    },
+    include: [
```
Comment (fontanierh) [inline front/lib/api/extract.ts:171]: 😱

File: front/lib/api/extract.ts [hunk 3]
```diff
@@ -154,3 +154,37 @@ export async function getExtractedEvents(
...
+      {
+        model: EventSchema,
+        as: "eventSchema",
+      },
+    ],
+  });
+
+  if (!event) {
+    throw "extracted_event_not_found";
```
Comment (fontanierh) [inline front/lib/api/extract.ts:180]: Should we throw `new Error("extracted_event_not_found")` instead ? Throwing raw strings means we don't get any stacktrace

File: front/lib/api/extract.ts [hunk 4]
```diff
@@ -154,3 +154,37 @@ export async function getExtractedEvents(
...
+  }
+
+  if (event.eventSchema.workspaceId !== owner.id) {
+    throw "extracted_event_auth_error";
```
Comment (fontanierh) [inline front/lib/api/extract.ts:184]: same here

File: front/lib/swr.ts [hunk 5]
```diff
@@ -14,11 +14,14 @@ import { GetMembersResponseBody } from "@app/pages/api/w/[wId]/members";
 import { GetProvidersResponseBody } from "@app/pages/api/w/[wId]/providers";
 import { GetChatSessionsResponseBody } from "@app/pages/api/w/[wId]/use/chats";
 import { GetEventSchemasResponseBody } from "@app/pages/api/w/[wId]/use/extract";
+import { GetExtractedEventsResponseBody } from "@app/pages/api/w/[wId]/use/extract/event/[id]";
 import { AppType } from "@app/types/app";
 import { DataSourceType } from "@app/types/data_source";
 import { RunRunType } from "@app/types/run";
 import { WorkspaceType } from "@app/types/user";
 
+import { EventSchema } from "./models";
```
Comment (fontanierh) [inline front/lib/swr.ts:23]: nit: relative import

**Review by fontanierh — COMMENTED** [review]
LGTM ! Just a couple minor comments

File: front/lib/api/extract.ts (hunk 3 above)
Comment (spolu) [inline front/lib/api/extract.ts:180]: This should not be a throw error I think. We can just return here no?

File: front/pages/api/w/[wId]/use/extract/event/[id]/index.ts [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+        auth,
+        parseInt(req.query.id as string)
+      );
+      return res.status(200).json({ events });
+
+    case "DELETE":
+      try {
+        await deleteExtractedEvent(auth, req.query.id as string);
+        res.status(200).end();
+      } catch (code) {
```
Comment (spolu) [inline front/pages/api/w/[wId]/use/extract/event/[id]/index.ts:69]: If you catch you have to use Result as per discussion a few days ago :)

**Review by spolu — CHANGES_REQUESTED** [review]

File: front/lib/api/extract.ts (hunk 1 above)
Comment (PopDaph) [reply front/lib/api/extract.ts:164]: Yeah I wondered if that was okay or not, they are all caught in the API route handler and an `apiError` for each, but it's true if you call this function from elsewhere we would throw error code, I'll make the change.

File: front/lib/api/extract.ts (hunk 3 above)
Comment (PopDaph) [reply front/lib/api/extract.ts:180]: Basically I did that to be able to return different `apiError` types depending on why we couldn't delete, but I get that we don't want to throw as is, I'll add a commit!

File: front/lib/api/extract.ts (hunk 1 above)
Comment (spolu) [reply front/lib/api/extract.ts:164]: Yes as per below. If you catch your error then you should use a Result :+1: with a `Result` you'll be able to embed an error code and a message and you'll be able to embed that in the final API response as you see fit and the order will be restored :)

File: front/lib/api/extract.ts (hunk 3 above)
Comment (spolu) [reply front/lib/api/extract.ts:180]: Fine to use a Result here as well :+1:

File: front/pages/w/[wId]/u/extract/[marker]/index.tsx [hunk 7]
```diff
@@ -61,9 +57,35 @@ export default function AppExtractEventsReadData({
...
   gaTrackingId,
 }: InferGetServerSidePropsType<typeof getServerSideProps>) {
+  const [isProcessing, setIsProcessing] = useState(false);
+  const { events, isEventsLoading } = useExtractedEvents(owner, schema.marker);
+
+  const handleDelete = async (eventId: ModelId) => {
+    if (window.confirm("Are you sure you want to delete?")) {
+      setIsProcessing(true);
+      const res = await fetch(
+        `/api/w/${owner.sId}/use/extract/${schema.marker}/events/${eventId}`,
```
Comment (spolu) [inline front/pages/w/[wId]/u/extract/[marker]/index.tsx:69]: marker should be encodeUriComponent

File: front/pages/w/[wId]/u/extract/[marker]/index.tsx [hunk 8]
```diff
@@ -61,9 +57,35 @@ export default function AppExtractEventsReadData({
...
+        {
+          method: "DELETE",
+        }
+      );
+      if (res.ok) {
+        await mutate(`/api/w/${owner.sId}/use/extract/${schema.marker}`);
```
Comment (spolu) [inline front/pages/w/[wId]/u/extract/[marker]/index.tsx:75]: same here

File: front/types/extract.ts [hunk 9]
```diff
@@ -8,6 +8,7 @@ export type EventSchemaType = {
   description?: string;
   status: EventSchemaStatus;
   properties: EventSchemaPropertyType[];
+  workspaceId: ModelId;
```
Comment (spolu) [inline front/types/extract.ts:11]: I'm not a big fan of that reverse dependency why do we need it?

File: front/lib/api/extract.ts [hunk 10]
```diff
@@ -124,21 +127,70 @@ export async function updateEventSchema(
...
-): Promise<ExtractedEventType[]> {
+export async function getExtractedEvent({
+  auth,
+  marker,
+  eventId,
+}: {
+  auth: Authenticator;
+  marker: string;
+  eventId: string;
+}) {
```
Comment (spolu) [inline front/lib/api/extract.ts:142]: We should return EventType here not the model?

File: front/pages/api/w/[wId]/use/extract/[marker]/events/index.ts [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+      api_error: {
+        type: "workspace_auth_error",
+        message:
+          "Only users of the current workspace can retrieve extracted events.",
+      },
+    });
+  }
+
+  const schema = await getEventSchema(auth, req.query.marker as string);
+  console.log(schema);
```
Comment (fontanierh) [inline front/pages/api/w/[wId]/use/extract/[marker]/events/index.ts:59]: nit: 🧹

File: front/lib/api/extract.ts [hunk 12]
```diff
@@ -124,21 +127,70 @@ export async function updateEventSchema(
...
-  schemaId: number
-): Promise<ExtractedEventType[]> {
+export async function getExtractedEvent({
+  auth,
+  marker,
+  eventId,
+}: {
+  auth: Authenticator;
+  marker: string;
+  eventId: string;
```
Comment (spolu) [inline front/lib/api/extract.ts:141]: Let's name it `eId` here out of consistency

**Review by fontanierh — DISMISSED** [review]
LGTM

File: front/pages/api/w/[wId]/use/extract/[marker]/events/[eId]/index.ts [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+        message: "The event was not found.",
+      },
+    });
+  }
+
+  switch (req.method) {
+    case "DELETE":
+      await deleteExtractedEvent({
+        auth,
+        eventId: req.query.eId as string,
```
Comment (spolu) [inline front/pages/api/w/[wId]/use/extract/[marker]/events/[eId]/index.ts:102]: should be event.sId ?

File: front/lib/api/extract.ts (hunk 10 above)
Comment (spolu) [inline front/lib/api/extract.ts:142]: Let's type this Promise<ExtractedEventType> ?

**Review by spolu — CHANGES_REQUESTED** [review]
Aweome, left a few request for change but nothing major!

File: front/lib/api/extract.ts (hunk 10 above)
Comment (PopDaph) [reply front/lib/api/extract.ts:142]: Indeed 😱

File: front/types/extract.ts (hunk 9 above)
Comment (PopDaph) [reply front/types/extract.ts]: I was using it on the api route to display a specific error, but actually it didn't even make any sense as I was checking something that could never ever happen 🤦🏻‍♀️ 
I added a commit to remove!

File: front/pages/api/w/[wId]/use/extract/[marker]/events/[eId]/index.ts (hunk 13 above)
Comment (PopDaph) [reply front/pages/api/w/[wId]/use/extract/[marker]/events/[eId]/index.ts]: Could be, for now it's `event.eventSchemaId` on the model. I can make the change but I would do it in another PR 🙇🏻‍♀️

**Review by spolu — APPROVED** [review]
LGTM

As per IRL

We probably want to move to having sId on Templates and Events. Enforce RESTfulness and remove reference to the tempalte on the type of the event (we should have all them handy when we work on them)

**Comment by PopDaph** [comment]
Thanks again for the thorough review, I learned a lot about our best practices with this one!
 I will merge as is and will introduce the changes as discussed in another PR 🙇🏻‍♀️

**Pull request merged**